lib/restforce/concerns/api.rb

```diff
@@ -360,9 +360,13 @@ def upsert(*args)
       # Raises an exception if an error is returned from Salesforce, including the 300
       # error returned if the external ID provided matches multiple records (in which
       # case the conflicting IDs can be found by looking at the response on the error)
-      def upsert!(sobject, field, attrs)
+      def upsert!(sobject, field, attrs_arg)
+        attrs = attrs_arg.dup
         external_id =
           extract_case_insensitive_string_or_symbol_key_from_hash!(attrs, field)
+        if field.to_s != "Id" && (external_id.nil? || external_id.strip.empty?)
```

Comment:
I think I'd prefer to leave this out for now, or at least to do it in a separate PR to keep things clear and simple.